stl/inc/xstring

```diff
@@ -2569,43 +2555,43 @@ private:
             return;
         }
 
-        // We need to check whether we have a misaligned SSO buffer because of the proxy in `_Container_base` (e.g. x86)
-        if constexpr (_Memcpy_val_offset % _Asan_granularity != 0) {
-            const uintptr_t _Alignment_offset = reinterpret_cast<uintptr_t>(_Ptr) & (_Asan_granularity - 1);
-            if (_Alignment_offset != 0 && _Capacity == _BUF_SIZE - 1) {
-                return;
-            }
-        }
+        // Note that `_Capacity`, `_Old_size`, and `_New_size` do not include the null terminator,
+        // so we need to add one back.
+        const void* const _Buf_last = _First + _Capacity + 1;
+        const void* const _Old_last = _First + _Old_size + 1;
+        const void* const _New_last = _First + _New_size + 1;
 
-        // Needs to consider the null terminator
-        const char* _First    = reinterpret_cast<const char*>(_Ptr);
-        const char* _End      = reinterpret_cast<const char*>(_Ptr + _Capacity + 1);
-        const char* _Old_last = reinterpret_cast<const char*>(_Ptr + _Old_size + 1);
-        const char* _New_last = reinterpret_cast<const char*>(_Ptr + _New_size + 1);
-        if constexpr (_Has_minimum_allocation_alignment_string<basic_string>) {
-            __sanitizer_annotate_contiguous_container(_First, _End, _Old_last, _New_last);
+        constexpr bool _Large_string_always_aligned =
+            (_Container_allocation_minimum_alignment<basic_string>) > _Asan_granularity;
+        constexpr bool _Small_string_always_aligned = _Memcpy_val_offset % _Asan_granularity == 0;
+        if constexpr (_Large_string_always_aligned && _Small_string_always_aligned) {
+            __sanitizer_annotate_contiguous_container(_First, _Buf_last, _Old_last, _New_last);
         } else {
-            const void* _Aligned_first = _Get_aligned_first(_First, _Capacity + 1);
+            const void* const _Aligned_first = _STD _Get_asan_aligned_first(_First, _Buf_last);
             if (!_Aligned_first) {
-                // There is no aligned address within the underlying buffer. Nothing to do
+                // There is no aligned address within the underlying buffer.
+                // This should be impossible, as the minimum allocation size for string is 16...

```

Comment:
This is assuming `char`; not all `basic_string`s are `string`.

Reply:
no it isn't; all `basic_string`s (as far as I know) have a small buffer of at least (byte) size `16`.

Reply:
Per:

https://github.com/microsoft/STL/blob/fb386c01083245a80db2666433235918c40d8363/stl/inc/xstring#L2273

`sizeof(value_type[_BUF_SIZE]) * sizeof(value_type)` is `16` iff `sizeof(value_type)` divides `16`, i.e., `sizeof(value_type)` is 1, 2, 4, 8, or 16. `sizeof(value_type[_BUF_SIZE]) >= 16` holds when `sizeof(_Ty)` is a power of two or greater than 15.

Reply:
right, exactly

Reply:
oh, I see what you're saying; I meant "the minimum allocation size for [all] string [types] is [at least] 16". Will fix.